File: modules/safespace-grpc/android/src/main/java/expo/modules/safespacegrpc/GrpcRuntimeClient.kt
package expo.modules.safespacegrpc

import Impactyn.Contracts.RuntimeService.V1.ImpactynContractsRuntimeServiceV1.CreateOrUpdateRequest
import Impactyn.Contracts.RuntimeService.V1.ImpactynContractsRuntimeServiceV1.DeleteRequest
import Impactyn.Contracts.RuntimeService.V1.ImpactynContractsRuntimeServiceV1.ExecuteRequest
import Impactyn.Contracts.RuntimeService.V1.ImpactynContractsRuntimeServiceV1.GetRequest
import Impactyn.Contracts.RuntimeService.V1.RuntimeServiceGrpc
import com.google.protobuf.ByteString
import io.grpc.ManagedChannel
import io.grpc.Metadata
import io.grpc.StatusRuntimeException
import io.grpc.okhttp.OkHttpChannelBuilder
import io.grpc.stub.MetadataUtils
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.asExecutor
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicReference

private const val AUTHORIZATION_HEADER_NAME = "authorization"
private const val CLIENT_VERSION_HEADER_NAME = "x-safespace-client-version"
private const val GRPC_DEADLINE_SECONDS = 60L

private val AUTHORIZATION_METADATA_KEY: Metadata.Key<String> =
  Metadata.Key.of(AUTHORIZATION_HEADER_NAME, Metadata.ASCII_STRING_MARSHALLER)
private val CLIENT_VERSION_METADATA_KEY: Metadata.Key<String> =
  Metadata.Key.of(CLIENT_VERSION_HEADER_NAME, Metadata.ASCII_STRING_MARSHALLER)

data class GrpcEndpoint(val host: String, val port: Int)

data class GrpcCallResult(val content: ByteArray, val headers: Map<String, String>)

class GrpcCallException(val statusCode: String, message: String) : Exception(message)

/**
 * Thin bridge to the RuntimeService gRPC endpoint. Every domain call in the backend
 * (login, feed, wallet, ...) goes through this one generic Get/Execute/CreateOrUpdate/Delete
 * service; the concrete request/response schema lives entirely in the `content` bytes, which
 * are encoded/decoded on the JS side.
 */
class GrpcRuntimeClient {
  private val endpointRef = AtomicReference(GrpcEndpoint(host = "staging.impactyn.io", port = 530))
  private val channelRef = AtomicReference<ManagedChannel?>(null)

  fun configure(host: String, port: Int) {
    val current = endpointRef.get()
    if (current.host == host && current.port == port) return
    endpointRef.set(GrpcEndpoint(host, port))
    channelRef.getAndSet(null)?.shutdownNow()
  }

  private fun channel(): ManagedChannel = channelRef.get() ?: synchronized(this) {
    channelRef.get() ?: run {
      val endpoint = endpointRef.get()
      val built = OkHttpChannelBuilder.forAddress(endpoint.host, endpoint.port)
        .useTransportSecurity()
        .executor(Dispatchers.IO.asExecutor())
        .build()
      channelRef.set(built)
      built
    }
  }

  private fun stub(token: String?, clientVersion: String) = run {
    val headers = Metadata()
    token?.let { headers.put(AUTHORIZATION_METADATA_KEY, it) }
    headers.put(CLIENT_VERSION_METADATA_KEY, clientVersion)

    RuntimeServiceGrpc.newBlockingStub(channel())
      .withDeadlineAfter(GRPC_DEADLINE_SECONDS, TimeUnit.SECONDS)
      .withInterceptors(MetadataUtils.newAttachHeadersInterceptor(headers))
  }

  fun get(
    apiVersion: String,
    namespace: String?,
    resource: String,
    name: String?,
    view: String?,
    token: String?,
    clientVersion: String,
  ): GrpcCallResult = runCatching {
    val request = GetRequest.newBuilder()
      .setApiVersion(apiVersion)
      .apply { namespace?.let { setNamespace(it) } }
      .setResource(resource)
      .apply { name?.let { setName(it) } }
      .setView(view ?: "")
      .build()
    val response = stub(token, clientVersion).get(request)
    GrpcCallResult(response.content.toByteArray(), response.headersMap)
  }.getOrElse { throw toGrpcCallException(it) }

  fun execute(
    apiVersion: String,
    namespace: String?,
    resource: String,
    name: String?,
    action: String?,
    payload: ByteArray?,
    token: String?,
    clientVersion: String,
  ): GrpcCallResult = runCatching {
    val request = ExecuteRequest.newBuilder()
      .setApiVersion(apiVersion)
      .apply { namespace?.let { setNamespace(it) } }
      .setResource(resource)
      .apply { name?.let { setName(it) } }
      .setAction(action ?: "")
      .apply { payload?.let { setContent(ByteString.copyFrom(it)) } }
      .build()
    val response = stub(token, clientVersion).execute(request)
    GrpcCallResult(response.content.toByteArray(), response.headersMap)
  }.getOrElse { throw toGrpcCallException(it) }

  fun createOrUpdate(
    apiVersion: String,
    namespace: String?,
    resource: String,
    name: String?,
    payload: ByteArray?,
    token: String?,
    clientVersion: String,
  ): GrpcCallResult = runCatching {
    val request = CreateOrUpdateRequest.newBuilder()
      .setApiVersion(apiVersion)
      .apply { namespace?.let { setNamespace(it) } }
      .setResource(resource)
      .apply { name?.let { setName(it) } }
      .apply { payload?.let { setContent(ByteString.copyFrom(it)) } }
      .build()
    val response = stub(token, clientVersion).createOrUpdate(request)
    GrpcCallResult(response.content.toByteArray(), response.headersMap)
  }.getOrElse { throw toGrpcCallException(it) }

  fun delete(
    apiVersion: String,
    namespace: String?,
    resource: String,
    name: String?,
    token: String?,
    clientVersion: String,
  ): GrpcCallResult = runCatching {
    val request = DeleteRequest.newBuilder()
      .setApiVersion(apiVersion)
      .apply { namespace?.let { setNamespace(it) } }
      .setResource(resource)
      .apply { name?.let { setName(it) } }
      .build()
    stub(token, clientVersion).delete(request)
    GrpcCallResult(ByteArray(0), emptyMap())
  }.getOrElse { throw toGrpcCallException(it) }

  private fun toGrpcCallException(error: Throwable): GrpcCallException {
    if (error is StatusRuntimeException) {
      return GrpcCallException(error.status.code.name, error.status.description ?: error.message.orEmpty())
    }
    return GrpcCallException("UNKNOWN", error.message ?: "Unknown gRPC error")
  }
}

File: modules/safespace-grpc/ios/GrpcRuntimeClient.swift
import Foundation
import GRPC
import NIOCore
import NIOPosix

private let authorizationHeaderName = "authorization"
private let clientVersionHeaderName = "x-safespace-client-version"
private let grpcDeadlineSeconds: Int64 = 60

struct GrpcEndpoint: Equatable {
  let host: String
  let port: Int
}

struct GrpcCallResult {
  let content: Data
  let headers: [String: String]
}

struct GrpcCallException: Error {
  let statusCode: String
  let message: String
}

/// Thin bridge to the RuntimeService gRPC endpoint. Mirrors GrpcRuntimeClient.kt:
/// every domain call goes through this one generic Get/Execute/CreateOrUpdate/Delete
/// service; the concrete request/response schema lives entirely in the `content`
/// bytes, which are encoded/decoded on the JS side.
final class GrpcRuntimeClient {
  private var endpoint = GrpcEndpoint(host: "staging.impactyn.io", port: 530)
  private var channel: GRPCChannel?
  private let group = MultiThreadedEventLoopGroup(numberOfThreads: 1)

  func configure(host: String, port: Int) {
    let newEndpoint = GrpcEndpoint(host: host, port: port)
    if newEndpoint == endpoint { return }
    endpoint = newEndpoint
    try? channel?.close().wait()
    channel = nil
  }

  private func getChannel() throws -> GRPCChannel {
    if let channel = channel { return channel }
    let built = try GRPCChannelPool.with(
      target: .host(endpoint.host, port: endpoint.port),
      transportSecurity: .tls(GRPCTLSConfiguration.makeClientDefault(compatibleWith: group)),
      eventLoopGroup: group
    )
    channel = built
    return built
  }

  private func callOptions(token: String?, clientVersion: String) -> CallOptions {
    var headers = HPACKHeaders()
    if let token = token {
      headers.add(name: authorizationHeaderName, value: token)
    }
    headers.add(name: clientVersionHeaderName, value: clientVersion)
    var options = CallOptions(customMetadata: headers)
    options.timeLimit = .timeout(.seconds(grpcDeadlineSeconds))
    return options
  }

  private func stub(token: String?, clientVersion: String) throws -> Impactyn_Contracts_RuntimeService_V1_RuntimeServiceClient {
    Impactyn_Contracts_RuntimeService_V1_RuntimeServiceClient(
      channel: try getChannel(),
      defaultCallOptions: callOptions(token: token, clientVersion: clientVersion)
    )
  }

  func get(
    apiVersion: String, namespace: String?, resource: String, name: String?,
    view: String?, token: String?, clientVersion: String
  ) async throws -> GrpcCallResult {
    do {
      let request = Impactyn_Contracts_RuntimeService_V1_GetRequest.with {
        $0.apiVersion = apiVersion
        if let namespace = namespace { $0.namespace = namespace }
        $0.resource = resource
        if let name = name { $0.name = name }
        $0.view = view ?? ""
      }
      let response = try await stub(token: token, clientVersion: clientVersion).get(request).response.get()
      return GrpcCallResult(content: response.content, headers: response.headers)
    } catch {
      throw Self.toGrpcCallException(error)
    }
  }

  func execute(
    apiVersion: String, namespace: String?, resource: String, name: String?,
    action: String?, payload: Data?, token: String?, clientVersion: String
  ) async throws -> GrpcCallResult {
    do {
      let request = Impactyn_Contracts_RuntimeService_V1_ExecuteRequest.with {
        $0.apiVersion = apiVersion
        if let namespace = namespace { $0.namespace = namespace }
        $0.resource = resource
        if let name = name { $0.name = name }
        $0.action = action ?? ""
        if let payload = payload { $0.content = payload }
      }
      let response = try await stub(token: token, clientVersion: clientVersion).execute(request).response.get()
      return GrpcCallResult(content: response.content, headers: response.headers)
    } catch {
      throw Self.toGrpcCallException(error)
    }
  }

  func createOrUpdate(
    apiVersion: String, namespace: String?, resource: String, name: String?,
    payload: Data?, token: String?, clientVersion: String
  ) async throws -> GrpcCallResult {
    do {
      let request = Impactyn_Contracts_RuntimeService_V1_CreateOrUpdateRequest.with {
        $0.apiVersion = apiVersion
        if let namespace = namespace { $0.namespace = namespace }
        $0.resource = resource
        if let name = name { $0.name = name }
        if let payload = payload { $0.content = payload }
      }
      let response = try await stub(token: token, clientVersion: clientVersion).createOrUpdate(request).response.get()
      return GrpcCallResult(content: response.content, headers: response.headers)
    } catch {
      throw Self.toGrpcCallException(error)
    }
  }

  func delete(
    apiVersion: String, namespace: String?, resource: String, name: String?,
    token: String?, clientVersion: String
  ) async throws -> GrpcCallResult {
    do {
      let request = Impactyn_Contracts_RuntimeService_V1_DeleteRequest.with {
        $0.apiVersion = apiVersion
        if let namespace = namespace { $0.namespace = namespace }
        $0.resource = resource
        if let name = name { $0.name = name }
      }
      _ = try await stub(token: token, clientVersion: clientVersion).delete(request).response.get()
      return GrpcCallResult(content: Data(), headers: [:])
    } catch {
      throw Self.toGrpcCallException(error)
    }
  }

  private static func toGrpcCallException(_ error: Error) -> GrpcCallException {
    if let status = error as? GRPCStatus {
      return GrpcCallException(statusCode: "\(status.code)", message: status.message ?? "Unknown gRPC error")
    }
    return GrpcCallException(statusCode: "UNKNOWN", message: error.localizedDescription)
  }
}
